File: node/be/api/steps/minicex.steps.js
'use strict';

const { When, Then } = require('@cucumber/cucumber');
const { MiniCex, ApiUnreachable } = require('../venues/minicex');

/**
 * API-tier steps for features/be-cex-api.feature. The account and action steps
 * (deposit, transfer, swap, bridge) are shared from be/db/steps/cex.steps.js;
 * here are the read-only calls and the assertions on the HTTP response itself.
 * A transport failure grades Blocked; a 4xx/5xx is an answer under test.
 */

const cex = new MiniCex();
const coin = (n) => Math.round(Number(n) * 1e8);
const acctOf = (world, alias) => world.accounts[alias];

async function act(world, fn) {
  if (world.sourceError) return;
  try { await fn(); } catch (err) { if (err instanceof ApiUnreachable) { world.sourceError = err.message; return; } throw err; }
}
function assert(world, description, cond, detail) {
  if (world.sourceError) { world.unobservable(description, 'the source could not be reached -- ' + world.sourceError); return; }
  world.check(description, cond, detail);
}

// ---------------------------------------------------------------- read-only actions

When('the assets are read', async function () { await act(this, async () => { this.api = await cex.get('/assets'); }); });
When('the pairs are read', async function () { await act(this, async () => { this.api = await cex.get('/pairs'); }); });
When('a quote of {float} {word} to {word} is requested', async function (amt, from, to) {
  await act(this, async () => { this.api = await cex.quote(from, to, coin(amt)); this.quote = this.api.body; });
});
When('a fresh account {string} is created', async function (alias) {
  await act(this, async () => { const handle = alias + '-' + Date.now().toString(36) + Math.random().toString(36).slice(2, 6); this.api = await cex.post('/accounts', { handle }); if (this.api.status === 201) this.accounts[alias] = this.api.body; });
});
When('an account is created with the same handle as {string}', async function (alias) {
  await act(this, async () => { this.api = await cex.post('/accounts', { handle: acctOf(this, alias).handle }); });
});
When('a deposit is attempted with no token', async function () {
  await act(this, async () => { this.api = await cex.post('/deposit', { asset: 'BTC', amount: coin(1) }); });
});

// ---------------------------------------------------------------- response assertions

Then('the response status is {int}', function (code) {
  assert(this, 'response status ' + code, this.api && this.api.status === code, this.api ? 'got ' + this.api.status + ' ' + (this.api.text || '').slice(0, 120) : 'no response');
});
Then('the response code is {string}', function (code) {
  assert(this, 'response code ' + code, this.api && this.api.body && this.api.body.code === code, this.api && this.api.body ? 'got ' + this.api.body.code : 'no body');
});
Then('the response has a token', function () {
  assert(this, 'response has token', this.api && this.api.body && typeof this.api.body.token === 'string' && this.api.body.token.length > 0, this.api && this.api.body ? JSON.stringify(this.api.body).slice(0, 80) : 'no body');
});
Then('the asset list includes {word}, {word}, {word}, {word}', function (a, b, c, d) {
  const have = new Set((this.api.body.assets || []).map((x) => x.symbol));
  const want = [a, b, c, d];
  assert(this, 'asset list includes ' + want.join(','), want.every((s) => have.has(s)), 'have ' + [...have].join(','));
});
Then('every listed asset has a name and a USD price', function () {
  const bad = (this.api.body.assets || []).filter((x) => !x.name || typeof x.usd_micro !== 'number');
  assert(this, 'assets have name + price', bad.length === 0 && (this.api.body.assets || []).length > 0, bad.length + ' incomplete');
});
Then('the pair list includes {word}\\/{word} and {word}\\/{word}', function (b1, q1, b2, q2) {
  const set = new Set((this.api.body.pairs || []).map((p) => p.base + '/' + p.quote));
  assert(this, 'pair list includes', set.has(b1 + '/' + q1) && set.has(b2 + '/' + q2), [...set].join(' '));
});
Then('no pair has the same asset on both legs', function () {
  const bad = (this.api.body.pairs || []).filter((p) => p.base === p.quote);
  assert(this, 'pairs have distinct legs', bad.length === 0, bad.length + ' bad');
});
Then('the quote net equals the gross minus the fee', function () {
  const q = this.quote;
  assert(this, 'quote net = gross - fee', q && q.to_amount === q.gross_amount - q.fee_amount, q ? `net ${q.to_amount}, gross ${q.gross_amount}, fee ${q.fee_amount}` : 'no quote');
});
Then('the quote fee matches the pair fee in basis points', function () {
  const q = this.quote;
  const expected = Math.floor((q.gross_amount * q.fee_bps) / 10000);
  assert(this, 'quote fee matches bps', q.fee_amount === expected, `fee ${q.fee_amount}, expected ${expected} at ${q.fee_bps}bps`);
});
Then('the API balance for {string} in {word} is {float} {word}', async function (alias, asset, amt, _a) {
  await act(this, async () => {
    const r = await cex.get('/balances/' + acctOf(this, alias).handle);
    const row = (r.body.balances || []).find((x) => x.asset === asset);
    const got = row ? row.amount : 0;
    assert(this, `API balance ${alias}/${asset}`, got === coin(amt), `got ${got}, want ${coin(amt)}`);
  });
});
Then('the API balance for {string} in {word} is above zero', async function (alias, asset) {
  await act(this, async () => {
    const r = await cex.get('/balances/' + acctOf(this, alias).handle);
    const row = (r.body.balances || []).find((x) => x.asset === asset);
    assert(this, `API balance ${alias}/${asset} > 0`, !!row && row.amount > 0, row ? ('got ' + row.amount) : 'no ' + asset + ' row');
  });
});
Then('the swap response output equals the quote for {float} {word} to {word}', async function (amt, from, to) {
  await act(this, async () => {
    const q = await cex.quote(from, to, coin(amt));
    assert(this, 'swap output == quote', this.api.body && this.api.body.to_amount === q.body.to_amount, `swap ${this.api.body && this.api.body.to_amount}, quote ${q.body.to_amount}`);
  });
});
Then('the second deposit was an idempotent replay', function () {
  assert(this, 'deposit idempotent replay', this.api && this.api.body && this.api.body.idempotent_replay === true, this.api && this.api.body ? JSON.stringify(this.api.body).slice(0, 80) : 'no body');
});
Then('the second bridge deposit was an idempotent replay', function () {
  assert(this, 'bridge deposit idempotent replay', this.api && this.api.body && this.api.body.idempotent_replay === true, this.api && this.api.body ? JSON.stringify(this.api.body).slice(0, 80) : 'no body');
});
Then('the bridge op is in status {string}', function (status) {
  assert(this, 'bridge op status ' + status, this.lastBridge && this.lastBridge.status === status, this.lastBridge ? 'got ' + this.lastBridge.status : 'no bridge op');
});

File: python/be/api/steps/minicex_steps.py
"""API-tier steps for be-cex-api.feature. Mirror of node/be/api/steps/minicex.steps.js.
The account and action steps are shared from be/db/steps/cex_steps.py; here are
the read-only calls and the assertions on the HTTP response itself.
"""

from __future__ import annotations

import random
import time

from pytest_bdd import parsers, then, when

from be.api.venues.minicex import ApiUnreachable, MiniCex

cex = MiniCex()


def coin(n):
    return round(float(n) * 1e8)


def acct(qa, alias):
    return qa.accounts[alias]


def act(qa, fn):
    if qa.source_error:
        return
    try:
        fn()
    except ApiUnreachable as err:
        qa.source_error = str(err)


def assert_(qa, description, cond, detail):
    if qa.source_error:
        qa.unobservable(description, "the source could not be reached -- " + qa.source_error)
        return
    qa.check(description, bool(cond), detail)


# ---------------------------------------------------------------- read-only actions

@when("the assets are read")
def read_assets(qa):
    act(qa, lambda: setattr(qa, "api", cex.get("/assets")))


@when("the pairs are read")
def read_pairs(qa):
    act(qa, lambda: setattr(qa, "api", cex.get("/pairs")))


@when(parsers.parse("a quote of {amt:f} {frm} to {to} is requested"))
def request_quote(qa, amt, frm, to):
    def go():
        qa.api = cex.quote(frm, to, coin(amt))
        qa.quote = qa.api["body"]
    act(qa, go)


@when(parsers.parse('a fresh account "{alias}" is created'))
def create_account_api(qa, alias):
    def go():
        handle = f"{alias}-{int(time.time()*1000):x}{random.randint(0, 65535):04x}"
        qa.api = cex.post("/accounts", {"handle": handle})
        if qa.api["status"] == 201:
            qa.accounts[alias] = qa.api["body"]
    act(qa, go)


@when(parsers.parse('an account is created with the same handle as "{alias}"'))
def create_duplicate(qa, alias):
    act(qa, lambda: setattr(qa, "api", cex.post("/accounts", {"handle": acct(qa, alias)["handle"]})))


@when("a deposit is attempted with no token")
def deposit_no_token(qa):
    act(qa, lambda: setattr(qa, "api", cex.post("/deposit", {"asset": "BTC", "amount": coin(1)})))


# ---------------------------------------------------------------- response assertions

@then(parsers.parse("the response status is {code:d}"))
def response_status(qa, code):
    assert_(qa, f"response status {code}", qa.api and qa.api["status"] == code, (f"got {qa.api['status']} {(qa.api['text'] or '')[:120]}" if qa.api else "no response"))


@then(parsers.parse('the response code is "{code}"'))
def response_code(qa, code):
    assert_(qa, f"response code {code}", qa.api and qa.api["body"] and qa.api["body"].get("code") == code, (f"got {qa.api['body'].get('code')}" if qa.api and qa.api["body"] else "no body"))


@then("the response has a token")
def response_token(qa):
    b = qa.api["body"] if qa.api else None
    assert_(qa, "response has token", b and isinstance(b.get("token"), str) and len(b["token"]) > 0, str(b)[:80] if b else "no body")


@then(parsers.parse("the asset list includes {a}, {b}, {c}, {d}"))
def asset_list_includes(qa, a, b, c, d):
    have = {x["symbol"] for x in (qa.api["body"]["assets"] or [])}
    want = [a, b, c, d]
    assert_(qa, "asset list includes " + ",".join(want), all(s in have for s in want), "have " + ",".join(sorted(have)))


@then("every listed asset has a name and a USD price")
def assets_complete(qa):
    lst = qa.api["body"]["assets"] or []
    bad = [x for x in lst if not x.get("name") or not isinstance(x.get("usd_micro"), int)]
    assert_(qa, "assets have name + price", len(bad) == 0 and len(lst) > 0, f"{len(bad)} incomplete")


@then(parsers.parse("the pair list includes {b1}/{q1} and {b2}/{q2}"))
def pair_list_includes(qa, b1, q1, b2, q2):
    s = {p["base"] + "/" + p["quote"] for p in (qa.api["body"]["pairs"] or [])}
    assert_(qa, "pair list includes", (b1 + "/" + q1) in s and (b2 + "/" + q2) in s, " ".join(sorted(s)))


@then("no pair has the same asset on both legs")
def pairs_distinct(qa):
    bad = [p for p in (qa.api["body"]["pairs"] or []) if p["base"] == p["quote"]]
    assert_(qa, "pairs have distinct legs", len(bad) == 0, f"{len(bad)} bad")


@then("the quote net equals the gross minus the fee")
def quote_net(qa):
    q = qa.quote
    assert_(qa, "quote net = gross - fee", q and q["to_amount"] == q["gross_amount"] - q["fee_amount"], (f"net {q['to_amount']}, gross {q['gross_amount']}, fee {q['fee_amount']}" if q else "no quote"))


@then("the quote fee matches the pair fee in basis points")
def quote_fee(qa):
    q = qa.quote
    expected = (q["gross_amount"] * q["fee_bps"]) // 10000
    assert_(qa, "quote fee matches bps", q["fee_amount"] == expected, f"fee {q['fee_amount']}, expected {expected} at {q['fee_bps']}bps")


@then(parsers.parse('the API balance for "{alias}" in {asset} is {amt:f} {a2}'))
def api_balance(qa, alias, asset, amt, a2):
    def go():
        r = cex.get("/balances/" + acct(qa, alias)["handle"])
        row = next((x for x in (r["body"]["balances"] or []) if x["asset"] == asset), None)
        got = row["amount"] if row else 0
        assert_(qa, f"API balance {alias}/{asset}", got == coin(amt), f"got {got}, want {coin(amt)}")
    act(qa, go)


@then(parsers.parse('the API balance for "{alias}" in {asset} is above zero'))
def api_balance_positive(qa, alias, asset):
    def go():
        r = cex.get("/balances/" + acct(qa, alias)["handle"])
        row = next((x for x in (r["body"]["balances"] or []) if x["asset"] == asset), None)
        assert_(qa, f"API balance {alias}/{asset} > 0", bool(row) and row["amount"] > 0, f"got {row['amount']}" if row else f"no {asset} row")
    act(qa, go)


@then(parsers.parse("the swap response output equals the quote for {amt:f} {frm} to {to}"))
def swap_output_matches(qa, amt, frm, to):
    def go():
        q = cex.quote(frm, to, coin(amt))
        assert_(qa, "swap output == quote", qa.api["body"] and qa.api["body"]["to_amount"] == q["body"]["to_amount"], f"swap {qa.api['body'].get('to_amount') if qa.api['body'] else None}, quote {q['body']['to_amount']}")
    act(qa, go)


@then("the second deposit was an idempotent replay")
def deposit_replay(qa):
    b = qa.api["body"] if qa.api else None
    assert_(qa, "deposit idempotent replay", b and b.get("idempotent_replay") is True, str(b)[:80] if b else "no body")


@then("the second bridge deposit was an idempotent replay")
def bridge_replay(qa):
    b = qa.api["body"] if qa.api else None
    assert_(qa, "bridge deposit idempotent replay", b and b.get("idempotent_replay") is True, str(b)[:80] if b else "no body")


@then(parsers.parse('the bridge op is in status "{status}"'))
def bridge_status_api(qa, status):
    assert_(qa, f"bridge op status {status}", qa.last_bridge and qa.last_bridge.get("status") == status, (f"got {qa.last_bridge['status']}" if qa.last_bridge else "no bridge op"))
